### DISCUSSION TITLE
make command palette show up on `cmd` + `p`

### DISCUSSION THREAD

**Discussion by o-az** [helix-editor/helix#10852] [Q&A] [answered]
Instead of (or in addition to) `space` + `shift` + `/` to pull up command palette, how do I also map `cmd` + `p` to trigger it?

**Comment by TornaxO7** [comment]
~~If I'm seeing it right in the [source code](https://github.com/helix-editor/helix/blob/master/helix-term/src/commands/typed.rs#L2490-L3111), it's currently not a typeable command, so it's (as far as I know) not directly possible to remap the command palette at the moment. Maybe you can achieve this by using macros (but I'm unsure about it).~~

I was wrong. You *can* open up the command palette with `cmd` + `p`!
Put that into your `config.toml`:
```toml
[keys.normal]
C-p = "command_palette"
```

**Reply by TornaxO7** [reply]
@o-az I updated my answer

**Reply by TornaxO7** [reply]
For more information, take a look into the [helix book](https://docs.helix-editor.com/remapping.html) :)

**Comment by xtrasmal** [comment, marked as answer]
@o-az 
Some keys aren't valid modifier keys(keys that can be pressed in combination with another). 
That's why folks use mapping tools like Karabiner to map the CMD into something else(like SHIFT+ALT+CTRL when pressed). 
Then you can map `C-S-A-p` which is like pressing `CMD+p`.

The helix docs state that:
**Ctrl**, **Shift** and **Alt** modifiers are encoded respectively with the prefixes **C-,** **S-** and **A-**. 

So @TornaxO7  if you do have it working, you must have mapped **Ctrl** to **Cmd** somewhere.

If somebody reads this and uses Karabiner Elements.. then the following might be handy for all your future keymapping dreams:

- MEH key (left_shift, left_option, left_control)
- HYPER key (left_shift, left_option, left_control, left_command)

**MEH: right_option to meh/right_option if pressed alone**
```json
{
    "description": "right_option to meh/right_option if alone",
    "manipulators": [
        {
            "from": {
                "key_code": "right_option",
                "modifiers": {
                    "optional": [
                        "any"
                    ]
                }
            },
            "to": [
                {
                    "key_code": "left_shift",
                    "lazy": true,
                    "modifiers": [
                        "left_option",
                        "left_control",
                        "left_shift"
                    ],
                    "repeat": false
                }
            ],
            "to_if_alone": [
                {
                    "key_code": "right_option"
                }
            ],
            "type": "basic"
        }
    ]
}
```


**HYPER: caps_lock to hyper/escape if pressed alone**
```json
{
    "description": "caps_lock to hyper/escape if alone",
    "manipulators": [
        {
            "from": {
                "key_code": "caps_lock",
                "modifiers": {
                    "optional": [
                        "any"
                    ]
                }
            },
            "to": [
                {
                    "key_code": "left_shift",
                    "lazy": true,
                    "modifiers": [
                        "left_command",
                        "left_option",
                        "left_control",
                        "left_shift"
                    ],
                    "repeat": false
                }
            ],
            "to_if_alone": [
                {
                    "key_code": "escape"
                }
            ],
            "type": "basic"
        },
        {
            "from": {
                "key_code": "escape"
            },
            "to": [
                {
                    "key_code": "caps_lock"
                }
            ],
            "type": "basic"
        }
    ]
}
```

@o-az if you want CMD+p to work, than first map some key.. like right_option to the CMD key just for your terminal. You can keep it for other applications. Then add other mappings, like the ones above to make it work.

BUT I would not advice you to do that.. because CMD is a special key for osx. Other apps might use it, like Spotlight or Raycast. In that case you won't be able to use those keys when working in the terminal.

**Reply by TornaxO7** [reply]
> So @TornaxO7 if you do have it working, you must have mapped Ctrl to Cmd somewhere.

oh.... I missread the word... I read `Ctrl` instead of `Cmd`... sorry
